src/PartitionService.ts

```diff
@@ -14,84 +14,73 @@
+                                    partitionStateVersion: number): void {
+        this.logger.debug('PartitionService',
+            'Handling new partition table with  partitionStateVersion: ' + partitionStateVersion);
+        if (!this.shouldBeApplied(connection, partitions, partitionStateVersion, this.partitionTable)) {
+            return;
         }
-        const ownerConnection = this.client.getClusterService().getOwnerConnection();
-        if (ownerConnection == null) {
-            return Promise.resolve();
-        }
-        const clientMessage: ClientMessage = GetPartitionsCodec.encodeRequest();
-
-        return this.client.getInvocationService()
-            .invokeOnConnection(ownerConnection, clientMessage)
-            .then((response: ClientMessage) => {
-                const receivedPartitionMap = GetPartitionsCodec.decodeResponse(response);
-                for (const partitionId in receivedPartitionMap) {
-                    this.partitionMap[partitionId] = receivedPartitionMap[partitionId];
-                }
-                this.partitionCount = Object.keys(this.partitionMap).length;
-            }).catch((e) => {
-                if (this.client.getLifecycleService().isRunning()) {
-                    this.logger.warn('PartitionService', 'Error while fetching cluster partition table from'
-                        + this.client.getClusterService().ownerUuid, e);
-                }
-            });
+        const newPartitions = this.convertToMap(partitions);
+        this.partitionTable.connection = connection;
+        this.partitionTable.partitionStateVersion = partitionStateVersion;
+        this.partitionTable.partitions = newPartitions;
     }
 
     /**
-     * Returns the {@link Address} of the node which owns given partition id.
      * @param partitionId
-     * @returns the address of the node.
+     * @return the owner of the partition or undefined if a partition is not assigned yet
      */
-    getAddressForPartition(partitionId: number): Address {
-        return this.partitionMap[partitionId];
+    public getPartitionOwner(partitionId: number): UUID {
+        return this.getPartitions().get(partitionId);
     }
 
     /**
      * Computes the partition id for a given key.
      * @param key
      * @returns the partition id.
+     * @throws ClientOfflineError if the partition table is not arrived yet.
      */
-    getPartitionId(key: any): number {
+    public getPartitionId(key: any): number {
+        if (this.partitionCount === 0) {
+            // Partition count can not be zero for the sync mode.
+            // On the sync mode, we are waiting for the first connection to be established.
+            // We are initializing the partition count with the value coming from the server with authentication.
+            // This exception is used only for async mode client.
+            throw new ClientOfflineError();
```

Comment:
This exception is thrown synchronously in the async start case. Say, it will be thrown on `BaseProxy.encodeInvokeOnKey()` call which happens in most map proxy methods. This behavior is implicit and surprising. We could improve it by rejecting the promise or waiting for the partition table initialization somehow (yet, the latter will contradict with Java client's behavior). WDYT?

Reply:
More context here. I've started discussion of this weird behavior on Slack.

Reply:
So, I take a look at the places in which we use this method. I believe `encodeInvokeOnKey` cannot throw sync error here because we need to first create proxy. To do so, client invokes the request on random connection which rejects with ClientOfflienError when async start set to true and there is no cluster connection yet. (ClientConnectionManager#isInvocationAllowed and does rejection). So, users shouldn't be able to reach places in which we use `encodeInvokeOnKey`. (We don't reset partition count anywhere, I belive Java client does this when changing clusters for the blue/green failover). But there was a problematic paths on `PartitionSpecificProxy`s (their constructor was throwing sync error for the above async start case). I fixed this issue in my last commit by making the call to their constructors after making a remote request (just like the Java client) and add a test. The other places that uses this method(near cache related code paths) calls this after client connects to cluster so there shouldn't be problem for them. Could you take another look ?

Reply:
Good findings. I've checked the commit and changes look good to me.